src/css-select-parse5-adapter.ts

```diff
@@ -0,0 +1,94 @@
+import {Adapter as CSSSelectAdapter} from 'css-select';
+import {Attribute, Element, Node, TreeAdapter} from 'parse5';
+const defaultTreeAdapter = require('parse5/lib/tree-adapters/default');
+export type Predicate = (node: Node) => boolean;
+export class Adapter implements CSSSelectAdapter<Node, Element> {
+  treeAdapter: TreeAdapter;
+  constructor(treeAdapter: TreeAdapter) {
+    this.treeAdapter = treeAdapter;
+  }
+  isTag(node: Node): node is Element {
+    return this.treeAdapter.isElementNode(node);
+  }
+  existsOne(test: Predicate, elems: Element[]): boolean {
+    return elems.some(test);
+  }
+  getAncestors(node: Node): Node[] {
+    const parentNode = this.hasParent(node) && this.getParent(node);
+    return parentNode ? [parentNode, ...this.getAncestors(parentNode)] : [];
+  }
+  getAttributeValue(elem: Element, name: string): string {
+    if (!this.isTag(elem)) {
+      return '';
+    }
+    const attrs = this.treeAdapter.getAttrList(elem);
+    for (const attr of attrs) {
+      if (attr.name === name) {
+        return attr.value;
+      }
+    }
+    return '';
+  }
+  getChildren(node: Node): Node[] {
+    return [...(this.treeAdapter.getChildNodes(node) || [])];
+  }
+  getName(elem: Element): string {
+    return this.treeAdapter.getTagName(elem);
+  }
+  getParent(node: Node): Node {
+    return this.treeAdapter.getParentNode(node);
+  }
+  getSiblings(node: Node): Node[] {
+    return this.getChildren(this.getParent(node));
+  }
+  getText(node: Node): string {
+    return this.treeAdapter.isTextNode(node) ?
+        this.treeAdapter.getTextNodeContent(node) :
+        this.getChildren(node).map((node: Node) => this.getText(node)).join('');
+  }
+  hasAttrib(elem: Element, name: string): boolean {
+    return this.isTag(elem) &&
+        this.treeAdapter.getAttrList(elem).some(
+            (attr: Attribute) => attr.name === name);
+  }
+  hasParent(node: Node): boolean {
+    return this.treeAdapter.isTextNode(node) ||
+        this.treeAdapter.isCommentNode(node) ||
+        this.treeAdapter.isElementNode(node) ||
+        this.treeAdapter.isDocumentTypeNode(node);
+  }
+  removeSubsets(nodes: Node[]): Node[] {
+    const filtered: Set<Node> = new Set(nodes);
+    for (const node of filtered) {
+      // Add node to the filtered set
+      if (this.getAncestors(node).some(
```

Comment:
Traversing up the tree will avoid creating ancestor arrays.

```ts
let ancestor = node;
while (ancestor = this.getParent(ancestor) !== undefined) {
  if (filtered.has(ancestor)) {
    ...
```

Reply:
Got rid of `getAncestors` and added a `findAncestor` traversal method:
```ts
  removeSubsets(nodes: Node[]): Node[] {
    const filtered: Set<Node> = new Set(nodes);
    for (const node of [...filtered]) {
      if (this.findAncestor((ancestor: Node) => filtered.has(ancestor), node)) {
        filtered.delete(node);
        continue;
      }
    }
    return [...filtered];
  }
  findAncestor(test: Predicate, node: Node): Node|undefined {
    do {
      node = this.getParent(node);
      if (node) {
        if (test(node)) {
          return node;
        }
      }
    } while (node);
    return undefined;
  }
```

Reply:
Oh, that brings up a question about which methods are implementing the interface and which are helpers. Might be good to note that with method naming or comments.

Reply:
Okay.

Reply:
I've decided to prefix any method with `_` if they are not part of the standard adapter interface.  So now it is `_findAncestor`.